python/src/trezorlib/cli/firmware.py

```diff
@@ -81,9 +127,57 @@ def validate_firmware(version, fw, expected_fingerprint=None):
         sys.exit(5)
 
 
+def check_device_match(
+    version: str,
+    fw: firmware.ParsedFirmware,
+    bootloader_onev2: bool,
+    trezor_major_version: int,
+) -> None:
+    """Validate if the device and firmware are compatible.
+
+    Prints error message and exits if the validation fails.
+    """
+    if (
+        bootloader_onev2
+        and version == firmware.FirmwareFormat.TREZOR_ONE
+        and not fw.embedded_onev2
+    ):
+        click.echo("Firmware is too old for your device. Aborting.")
+        sys.exit(3)
+    elif not bootloader_onev2 and version == firmware.FirmwareFormat.TREZOR_ONE_V2:
+        click.echo("You need to upgrade to bootloader 1.8.0 first.")
+        sys.exit(3)
+
+    if trezor_major_version not in ALLOWED_FIRMWARE_FORMATS:
+        click.echo("trezorctl doesn't know your device version. Aborting.")
+        sys.exit(3)
+    elif version not in ALLOWED_FIRMWARE_FORMATS[trezor_major_version]:
+        click.echo("Firmware does not match your device, aborting.")
+        sys.exit(3)
+
+
 def find_best_firmware_version(
-    bootloader_version, requested_version=None, beta=False, bitcoin_only=False
-):
+    client: TrezorClient,
+    version: str = None,
+    beta: bool = False,
+    bitcoin_only: bool = False
+) -> tuple:
+    """Get the url from which to download the firmware and its expected fingerprint."""
+    f = client.features
+    model = f.model or "1"
+    bootloader_version = [f.major_version, f.minor_version, f.patch_version]
```

Comment:
I am not sure what are the implications here, is it only the name of the variable that should change?

Theoretically, this X.X.X version does not change depending on being in BL mode or not. When running `trezorctl get-features`, it contains the same three version numbers in both cases.

We may use `f.major_version` instead of `bootloader_version[0]` and only initialize the whole list when comparing it with `release["min_bootloader_version"]`, it is its only full usage.

Reply:
> Theoretically, this X.X.X version does not change depending on being in BL mode

This is an incorrect assumption. Bootloader has its own version. And it is important for the selection algorithm below - namely that certain firmware versions (1.7.0 and up) require certain minimum bootloader version (1.6.0). The releases.json contains `min_bootloader_version` field which specifies that.

So if the device is _not_ in bootloader mode, we need to look at the firmware version, lookup its `bootloader_version` in the json, and then use that for comparison with `min_bootloader_version`.
...alternately, I just noticed there is also a `min_firmware_version`, which seems to match the result of this lookup. So i.e., I didn't verify this, but it should be possible to write "if is_bootloader: compare version to min_bootloader_version; else: compare version to min_firmware_version"